staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -486,6 +482,19 @@ func (cnc *CloudNodeController) syncNode(ctx context.Context, nodeName string) e
 			modify(newNode)
 		}
 
+		// spec.ProviderID is required for multiple controllers, like loadbalancers, so we should not
+		// untaint the node until is set. Once it is set, the field is immutable, so no need to reconcile.
+		// We only set this value during initialization and is never reconciled, so if for some reason
+		// we are not able to set it, the instance will never be able to acquire it.
+		// Before external cloud providers were enabled by default, the field was set by the kubelet, and the
+		// node was created with the value.
+		// xref: https://issues.k8s.io/123024
+		if !utilfeature.DefaultFeatureGate.Enabled(features.OptionalProviderID) {
+			if newNode.Spec.ProviderID == "" {
```

Comment:
Not against this but for this to have happened I assume that we failed to get the ProviderID at the earlier `cnc.getProviderID(ctx, copyNode)` call. Could we have exited earlier. Is there any chance we get a provider id earlier  but its not set in the newNode? If so should we fix it with the retrieved provider id?

Reply:
That is the part that I find hard to understand, because `getProviderID` can return an empty value if InstanscesV2 

https://github.com/kubernetes/kubernetes/blob/210378e918a225611719772ab728c127a4e51cd6/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L603-L611

but then if there is no metadata it is not reconciled

https://github.com/kubernetes/kubernetes/blob/210378e918a225611719772ab728c127a4e51cd6/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L442-L446


@cheftako my question is, a cloud provider MUST always return instance metadata?

If that is the case we need to error and retry, if that is not the case we can not return error and retry as it will hotloop.


/hold

waiting for this answer

Reply:
though I checked on the failing logs and there are no logs about `Skip sync` and thousands of logs about missing providerID

```
 grep -ri ProviderID cloud-controller-manager* | wc
 160048 2479273 36907480
```

```
rep -ri skip cloud-controller-manager*
cloud-controller-manager (1).log:      --authentication-skip-lookup                        If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
cloud-controller-manager (1).log:      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server. (default [/healthz,/readyz,/livez])
cloud-controller-manager (1).log:      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io. This is optional. If empty, all requests not skipped by authorization are forbidden.
cloud-controller-manager (1).log:      --authentication-skip-lookup                        If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
cloud-controller-manager (1).log:      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server. (default [/healthz,/readyz,/livez])
cloud-controller-manager (1).log:      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io. This is optional. If empty, all requests not skipped by authorization are forbidden.
cloud-controller-manager (1).log:W0220 17:06:21.837595      11 controllermanager.go:325] Skipping "node-route-controller"
cloud-controller-manager.log:      --authentication-skip-lookup                        If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
cloud-controller-manager.log:      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server. (default [/healthz,/readyz,/livez])
cloud-controller-manager.log:      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io. This is optional. If empty, all requests not skipped by authorization are forbidden.
cloud-controller-manager.log:      --authentication-skip-lookup                        If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
cloud-controller-manager.log:      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server. (default [/healthz,/readyz,/livez])
cloud-controller-manager.log:      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io. This is optional. If empty, all requests not skipped by authorization are forbidden.
cloud-controller-manager.log:W0214 17:07:01.528566      11 controllermanager.go:325] Skipping "node-route-controller"
```